// @mui-treasury/components/esm/flex/RowToColumn.js
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React from 'react';
import cx from 'clsx';
import makeStyles from '@material-ui/core/styles/makeStyles';
import { Provider, useGapLookup, BreakpointProvider, useBreakpointLookup } from './core';
import Item from './Item';
import { gapToCss, getLowerMediaQuery } from './utils';

var styles = function styles(theme) {
  var breakpoints = theme.breakpoints;
  return {
    root: function root(_ref) {
      var _ref3;

      var at = _ref.at,
          gap = _ref.gap,
          columnStyle = _ref.columnStyle,
          rowStyle = _ref.rowStyle,
          columnReversed = _ref.columnReversed;
      return _ref3 = {
        display: 'flex'
      }, _defineProperty(_ref3, breakpoints.up(at), _extends(_extends({}, columnStyle), {}, {
        flexDirection: columnReversed ? 'column-reverse' : 'column'
      }, gapToCss(theme)(gap, function (value) {
        return _defineProperty({}, "& > *:not(:".concat(columnReversed ? 'last' : 'first', "-child)"), {
          paddingTop: value
        });
      }))), _defineProperty(_ref3, getLowerMediaQuery(theme, at), _extends(_extends({
        alignItems: 'center'
      }, rowStyle), gapToCss(theme)(gap, function (value) {
        return {
          '& > *:not(:first-child)': {
            paddingLeft: value
          }
        };
      }))), _ref3;
    }
  };
};

var useStyles = makeStyles(styles, {
  name: 'FlexRowToColumn'
});

var RowToColumn = function RowToColumn(_ref4) {
  var className = _ref4.className,
      children = _ref4.children,
      gap = _ref4.gap,
      at = _ref4.at,
      columnReversed = _ref4.columnReversed,
      columnStyle = _ref4.columnStyle,
      rowStyle = _ref4.rowStyle,
      props = _objectWithoutProperties(_ref4, ["className", "children", "gap", "at", "columnReversed", "columnStyle", "rowStyle"]);

  var _useBreakpointLookup = useBreakpointLookup(at),
      calculatedAt = _useBreakpointLookup.calculatedAt;

  var _useGapLookup = useGapLookup(gap),
      calculatedGap = _useGapLookup.calculatedGap,
      itemProps = _useGapLookup.itemProps;

  var styles = useStyles({
    at: calculatedAt,
    columnReversed: columnReversed,
    columnStyle: columnStyle,
    rowStyle: rowStyle,
    gap: calculatedGap
  });
  return /*#__PURE__*/React.createElement(Item, _extends({
    className: cx('FlexRowToColumn', styles.root, className)
  }, itemProps, props), /*#__PURE__*/React.createElement(Provider, {
    flexDirection: 'row-column',
    gap: calculatedGap
  }, /*#__PURE__*/React.createElement(BreakpointProvider, {
    at: calculatedAt
  }, children)));
};

export default RowToColumn;
